src/file.c

```diff
@@ -0,0 +1,79 @@
+/* SPDX-License-Identifier: LGPL-3.0-or-later */
+/** @file file.c
+ *
+ * @author Juuso Alasuutari
+ */
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+
+#include "file.h"
+
+struct file_in
+file_read (char const *path)
+{
+	struct file_in ret = {0};
+
+	do {
+		FILE *fp = fopen(path, "rbem");
+		if (!fp) {
+			ret.ec[0] = errno;
+			break;
+		}
+
+		do {
+			struct stat s = {0};
+			ret.ec[0] = fileno(fp);
+
+			ret.ec[0] = ret.ec[0] < 0 || fstat(ret.ec[0], &s);
+			if (ret.ec[0]) {
+				ret.ec[0] = errno;
+				break;
+			}
+
+			if ((s.st_mode & S_IFMT) != S_IFREG) {
+				ret.ec[0] = EINVAL;
+				break;
+			}
+
+			if (s.st_size < 0) {
+				ret.ec[0] = EIO;
+				break;
+			}
+
+			ret.size = (size_t)s.st_size;
+			ret.data = malloc(ret.size + 1U);
+			if (!ret.data) {
+				ret.size = 0;
+				ret.ec[0] = errno;
+				break;
+			}
+
+			if (fread(ret.data, 1, ret.size, fp) != ret.size) {
+				ret.size = 0;
+				ret.ec[0] = errno;
+				free(ret.data);
+				ret.data = nullptr;
+				break;
+			}
+
+			ret.data[ret.size] = 0;
+		} while (0);
+
+		fclose(fp);
+		fp = nullptr;
```

Comment:
You have a point, kind of. It should be cast to `void` explicitly. But if the read succeeded we don't care. There's nothing we can do anyway. From `man fclose`:

```
Upon successful completion, 0 is returned.  Otherwise, EOF is
returned and errno is set to indicate the error.  In either case,
any further access (including another call to fclose()) to the
stream results in undefined behavior.
```

`struct file_in::ec[0]` is unavailable once we use the enclosing union to store the buffer size, and from a practical standpoint a failed `fclose()` does not affect the caller.